### PR TITLE
225 rating and comments

### PR DESCRIPTION
**Pull request by christinasugiono**
Only a user that visited the airport can contribute to `rating` and `comments`

image

clsoes #225

### CODE REVIEW

**Review by alexstan67 — COMMENTED** [review]
Hi @christinasugiono, here are some comments:
- It seems not possible to add a review on our homebase airport
- Once the review is done, and the page is loaded, we have a eager loading issue
- If your give a rating of 3 for example, than in the comment, you I guess you should also display 2 "dark" stars as well, no?

File: app/assets/stylesheets/components/_modals.scss
Comment (alexstan67) [inline app/assets/stylesheets/components/_modals.scss:1]: Very good!

File: app/controllers/airports_controller.rb [hunk 1]
```diff
@@ -21,6 +21,20 @@ def show
     @pilots = @airport.pilots.includes(user: :picture_attachment)
     @visited_airport = VisitedAirport.find_by(airport: @airport,
                                               pilot: current_user.pilot)
+    @rating = Rating.new
+
+    order_by = 'CASE WHEN comment IS NULL THEN 1 ELSE 0 END,
+                ratings.created_at DESC'
+    @ratings = @airport.ratings
+                       .includes(visited_airport: {
```
Comment (alexstan67) [inline app/controllers/airports_controller.rb:29]: I have an eager loading alert here, but not confident enough to remove it:
````bash
[2024-05-21 15:00:50 +0200] user: alex
GET /airports/2563
AVOID eager loading detected
  Pilot => [:user, :airport]
  Remove from your query: .includes([:user, :airport])
Call stack
  /home/alex/github/sky-unlimited/wef_2023/app/views/ratings/_ratings.html.erb:34:in `_app_views_ratings__ratings_html_erb___1299818406138423748_41100'
  /home/alex/github/sky-unlimited/wef_2023/app/views/airports/show.html.erb:313:in `block in _app_views_airports_show_html_erb___388856586667220054_40680'
  /home/alex/github/sky-unlimited/wef_2023/app/views/airports/show.html.erb:312:in `_app_views_airports_show_html_erb___388856586667220054_40680'
  /home/alex/github/sky-unlimited/wef_2023/app/controllers/application_controller.rb:19:in `switch_locale'


[2024-05-21 15:00:50 +0200] user: alex
GET /airports/2563
AVOID eager loading detected
  Rating => [:visited_airport]
  Remove from your query: .includes([:visited_airport])
Call stack
  /home/alex/github/sky-unlimited/wef_2023/app/views/ratings/_ratings.html.erb:34:in `_app_views_ratings__ratings_html_erb___1299818406138423748_41100'
  /home/alex/github/sky-unlimited/wef_2023/app/views/airports/show.html.erb:313:in `block in _app_views_airports_show_html_erb___388856586667220054_40680'
  /home/alex/github/sky-unlimited/wef_2023/app/views/airports/show.html.erb:312:in `_app_views_airports_show_html_erb___388856586667220054_40680'
  /home/alex/github/sky-unlimited/wef_2023/app/controllers/application_controller.rb:19:in `switch_locale'
````

File: app/controllers/ratings_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+class RatingsController < ApplicationController
+  def create
+    @visited_airport = VisitedAirport.find(params[:visited_airport_id])
+    @rating = Rating.new(rating_params)
+    @rating.visited_airport = @visited_airport
+    @source = params[:source]
+    if @rating.save
+      @rating = Rating.new
+      order_by = 'CASE WHEN comment IS NULL THEN 1 ELSE 0 END, created_at DESC'
+      @ratings = @visited_airport.airport.ratings
```
Comment (alexstan67) [inline app/controllers/ratings_controller.rb:11]: I don't know if there is also an eager loading problem here...

File: app/views/airports/show.html.erb [hunk 3]
```diff
@@ -3,20 +3,17 @@
 <div class="container pt-30">
   <div id="airport-header">
     <div>
-      <h1><%= NormalizeCountry(@airport.country.code, :to => :emoji) %> <br> <%= "#{@airport.name} (#{@airport.icao})" %></h1>
+      <h1><%= NormalizeCountry(@airport.country.code, :to => :emoji) %> <br> <%= @airport.display_name %></h1>
```
Comment (alexstan67) [inline app/views/airports/show.html.erb:6]: Seems clever to proceed so, it brings consistency through all application. But where is defined the method `display_name` ?

File: app/views/preferences/edit.html.erb [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <div class="container-sm">
   <%= render "weather_info" %>
   <div class="form-wrapper">
-    <%= form_for @preference, method: :patch, data: { turbo: false } do |f| %>
+    <%= form_for @preference, url: edit_preferences_path, data: { turbo: false } do |f| %>
```
Comment (alexstan67) [inline app/views/preferences/edit.html.erb:4]: I have this error:
No route matches [PATCH] "/preferences/edit"

Comment (alexstan67) [reply app/views/preferences/edit.html.erb:4]: It also means our controller tests are not good enough for this one :smile:

File: config/environment.rb
Comment (alexstan67) [inline config/environment.rb:1]: Should it not be present in `config/environments/production.rb` ?

File: config/locales/fr.yml [hunk 5]
```diff
@@ -255,6 +255,15 @@ fr:
       pois: Points d'intérêts SUR/PRÈS aérodrome
       upcoming_events: Évènements à venir
       no_events: Pas d'évènements
+      ratings: Évaluation
+      new_rating: Nouvelle Évaluation
+      no_ratings: Pas d'évaluations
+      overall_rating: Évaluation globale
+      already_rated: Vous avez déjà évalué cet aéroport avec %{score} ⭐️ en %{date}
+      reviews: Reviewaire
```
Comment (alexstan67) [inline config/locales/fr.yml:263]:
```suggestion
      reviews: Revues
```

File: config/locales/fr.yml [hunk 6]
```diff
@@ -255,6 +255,15 @@ fr:
       pois: Points d'intérêts SUR/PRÈS aérodrome
       upcoming_events: Évènements à venir
       no_events: Pas d'évènements
+      ratings: Évaluation
```
Comment (alexstan67) [inline config/locales/fr.yml:258]:
```suggestion
      ratings: Évaluations
```

File: config/locales/fr.yml [hunk 7]
```diff
@@ -255,6 +255,15 @@ fr:
...
+      new_rating: Nouvelle Évaluation
+      no_ratings: Pas d'évaluations
+      overall_rating: Évaluation globale
+      already_rated: Vous avez déjà évalué cet aéroport avec %{score} ⭐️ en %{date}
+      reviews: Reviewaire
+      new_review: Nouveau Reviewaire
```
Comment (alexstan67) [inline config/locales/fr.yml:264]:
```suggestion
      new_review: Nouvelle évaluation
```

File: config/locales/fr.yml [hunk 8]
```diff
@@ -255,6 +255,15 @@ fr:
...
+      no_reviews: Pas de reviewaires. Soyez le premier!
```
Comment (alexstan67) [inline config/locales/fr.yml:265]:
```suggestion
      no_reviews: Pas d'évaluations. Soyez le premier!
```

File: config/locales/fr.yml [hunk 9]
```diff
@@ -255,6 +255,15 @@ fr:
...
+      more_reviews: Afficher tous les %{count} reviewaires
```
Comment (alexstan67) [inline config/locales/fr.yml:266]:
```suggestion
      more_reviews: Afficher toutes les %{count} évaluations
```

File: config/locales/fr.yml [hunk 10]
```diff
@@ -295,3 +304,6 @@ fr:
   languages:
     en: Anglais
     fr: Français
+  reviews:
+    title: Reviewaires pour %{airport}
```
Comment (alexstan67) [inline config/locales/fr.yml:308]:
```suggestion
    title: Evaluations pour %{airport}
```

File: test/fixtures/ratings.yml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+valid_rating:
+  visited_airport: valid_visited_airport
+  score: 3
+  active: true
+
+invalid_rating:
```
Comment (alexstan67) [inline test/fixtures/ratings.yml:8]: Maybe let's discuss it, I used to make the same invalid tests, but it has no sense I think to insert a bad record in database... I think it should be tested with a creation having n bad values...

File: test/controllers/ratings_controller_test.rb
Comment (alexstan67) [inline test/controllers/ratings_controller_test.rb:1]: I think we should write 1 or 2 tests, WDYT? At least is it accessible when not been there? Accessible when visited?

File: config/routes.rb
Comment (alexstan67) [inline config/routes.rb:1]: Should we add an entry in the menu only in `development` environment? I didn't know this, it's interesting, but I don't understand how it works :disappointed:

File: app/views/ratings/create.turbo_stream.erb
Comment (alexstan67) [inline app/views/ratings/create.turbo_stream.erb:1]: Amazing, can you explain me how it works?

**Review by alexstan67 — CHANGES_REQUESTED** [review]
Hi @christinasugiono, I took time to review. You did crazy work, I would not have been able to implement this I think... Good job, but let's discuss some points tomorrow during meeting!

File: app/controllers/airports_controller.rb (hunk 1 above)
Comment (christinasugiono) [reply app/controllers/airports_controller.rb:29]: I cannot 100% fix this, one scenario still give a warning which is when the user creates the first comment for the airport. After there's another comment given from another user, the warning is gone.

I tried following the suggestion from bullet gem, but in the suggestion is circling.. 😮‍💨  First it asked me to add the `.includes` then after I add, it asked me to remove it....

File: app/views/airports/show.html.erb (hunk 3 above)
Comment (christinasugiono) [reply app/views/airports/show.html.erb:6]: It's a method from `airport.rb` model, I feel like I repeated this display name to many times, better create a method for it :D

File: app/views/preferences/edit.html.erb (hunk 4 above)
Comment (christinasugiono) [reply app/views/preferences/edit.html.erb:4]: image

I think we need to check this together then, cause on my side its okay....

File: config/routes.rb
Comment (christinasugiono) [reply config/routes.rb:1]: Good idea! I added it in the menu for `admin` and `development` env

Screenshot 2024-05-22 at 13 32 52

File: config/environment.rb
Comment (christinasugiono) [reply config/environment.rb:1]: Maybe, but it was there already before, so was not sure if I should move it 😅

**Review by alexstan67 — APPROVED** [review]

**Pull request merged**